File: golang/baseball/baseball_service.go
package baseball

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

func Execute(baseUrl string, query string, now time.Time) string {
	team := fetchMlbTeam(baseUrl, query, now)
	if team == nil {
		return "Not Found"
	}

	// Fetch roster asynchronously
	rosterChan := make(chan *mlbRoster)
	go func() {
		roster := fetchMlbRoster(baseUrl, team.Id)
		rosterChan <- roster
	}()

	// Fetch venue asynchronously
	venueChan := make(chan *mlbVenue)
	go func() {
		venue := fetchMlbVenueById(baseUrl, team.Venue.Id)
		venueChan <- venue
	}()

	// Wait for results from both channels
	roster := <-rosterChan
	close(rosterChan)

	venue := <-venueChan
	close(venueChan)

	return printRoster(*roster, *team, *venue)
}

func fetchMlbRoster(baseUrl string, teamId int) *mlbRoster {
	apiUrl := fmt.Sprintf("%s/api/v1/teams/%d/roster", baseUrl, teamId)
	req, _ := http.NewRequest("GET", apiUrl, nil)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return &mlbRoster{}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		fmt.Println("Response Status: ", resp.Status)
		return &mlbRoster{}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response body:", err)
		return &mlbRoster{}
	}

	var data mlbRoster
	if err := json.Unmarshal(body, &data); err != nil {
		fmt.Println("Body: " + string(body))
		fmt.Println("Error unmarshaling JSON: ", err)
		return &mlbRoster{}
	}

	return &data
}

func fetchMlbTeam(baseUrl string, query string, now time.Time) *mlbTeam {
	_, err := strconv.Atoi(query)
	isInt := err == nil
	if isInt {
		return fetchMlbTeamById(baseUrl, query)
	} else {
		return fetchMlbTeamByName(baseUrl, query, now)
	}
}

func fetchMlbTeamById(baseUrl string, id string) *mlbTeam {
	apiUrl := baseUrl + "/api/v1/teams/" + id
	req, _ := http.NewRequest("GET", apiUrl, nil)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		fmt.Println("Could not find team with id " + id)
		return nil
	} else if resp.StatusCode != http.StatusOK {
		fmt.Println("Response Status: ", resp.Status)
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response body:", err)
		return nil
	}

	var data mlbTeams
	if err := json.Unmarshal(body, &data); err != nil {
		fmt.Println("Body: " + string(body))
		fmt.Println("Error unmarshaling JSON: ", err)
		return nil
	}

	if len(data.Teams) != 1 {
		fmt.Println("Could not find team with id " + id)
		return nil
	}

	return &data.Teams[0]
}

func fetchMlbTeamByName(baseUrl string, teamName string, now time.Time) *mlbTeam {
	apiUrl := baseUrl + "/api/v1/teams/"
	req, _ := http.NewRequest("GET", apiUrl, nil)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")

	query := req.URL.Query()
	query.Add("season", fmt.Sprintf("%d", now.Year()))
	query.Add("sportIds", "1")
	req.URL.RawQuery = query.Encode()

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		fmt.Println("Could not find team with name " + teamName)
		return nil
	} else if resp.StatusCode != http.StatusOK {
		fmt.Println("Response Status: ", resp.Status)
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response body:", err)
		return nil
	}

	var data mlbTeams
	if err := json.Unmarshal(body, &data); err != nil {
		fmt.Println("Body: " + string(body))
		fmt.Println("Error unmarshaling JSON: ", err)
		return nil
	}

	for _, team := range data.Teams {
		if strings.Contains(strings.ToLower(team.Name), strings.ToLower(teamName)) {
			return &team
		}
	}

	fmt.Println("Could not find team with name " + teamName)
	return nil
}

func fetchMlbVenueById(baseUrl string, venueId int) *mlbVenue {
	apiUrl := fmt.Sprintf("%s/api/v1/venues/%d", baseUrl, venueId)
	req, _ := http.NewRequest("GET", apiUrl, nil)
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error sending request:", err)
		return nil
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		fmt.Printf("Could not find venue with id %d\n", venueId)
		return nil
	} else if resp.StatusCode != http.StatusOK {
		fmt.Printf("Response Status for venueId: %d: %s\n", venueId, resp.Status)
		return nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Error reading response body:", err)
		return nil
	}

	var data mlbVenues
	if err := json.Unmarshal(body, &data); err != nil {
		fmt.Println("Body: " + string(body))
		fmt.Println("Error unmarshaling JSON: ", err)
		return nil
	}

	if len(data.Venues) != 1 {
		fmt.Printf("Could not find venue with id %d\n", venueId)
		return nil
	}

	return &data.Venues[0]
}

func printRoster(roster mlbRoster, team mlbTeam, venue mlbVenue) string {
	var csvContent bytes.Buffer
	writer := csv.NewWriter(&csvContent)

	err := writer.Write([]string{"Team", "Jersey", "Name", "Position", "Home Stadium"})
	if err != nil {
		fmt.Println("Error flushing CSV writer:", err)
	}

	// Write the data to the CSV buffer
	for _, player := range roster.Roster {
		row := []string{
			team.Name,
			player.JerseyNumber,
			player.Person.FullName,
			player.Position.Abbreviation,
			venue.Name,
		}
		if err := writer.Write(row); err != nil {
			fmt.Println("Error writing record to CSV:", err)
		}
	}

	writer.Flush()
	if err := writer.Error(); err != nil {
		fmt.Println("Error flushing CSV writer:", err)
	}

	return csvContent.String()
}

type mlbPerson struct {
	Id       int    `json:"id"`
	FullName string `json:"fullName"`
}

type mlbPlayer struct {
	Person       mlbPerson   `json:"person"`
	JerseyNumber string      `json:"jerseyNumber"`
	Position     mlbPosition `json:"position"`
	Status       mlbStatus   `json:"status"`
}

type mlbPosition struct {
	Name         string `json:"name"`
	PositionType string `json:"type"`
	Abbreviation string `json:"abbreviation"`
}

type mlbRoster struct {
	Roster []mlbPlayer `json:"roster"`
}

type mlbStatus struct {
	Description string `json:"description"`
}

type mlbTeams struct {
	Teams []mlbTeam `json:"teams"`
}

type mlbTeam struct {
	Id           int    `json:"id"`
	Active       bool   `json:"active"`
	Name         string `json:"name"`
	LocationName string `json:"locationName"`
	TeamName     string `json:"teamname"`
	Venue        id     `json:"venue"`
}

type mlbVenues struct {
	Venues []mlbVenue `json:"venues"`
}

type mlbVenue struct {
	Id     int    `json:"id"`
	Name   string `json:"name"`
	Active bool   `json:"active"`
}

type id struct {
	Id int `json:"id"`
}

File: python/src/baseball/baseball_service.py
import asyncio
import csv
import datetime
import io
import json
import logging.config
import requests

HEADERS = {
  'Accept': 'application/json',
  'Content-Type': 'application/json',
  'Cache-Control': 'no-cache, no-store, must-revalidate',
  'Pragma': 'no-cache',
}

class BaseballService:

  def __init__(self, base_url='https://statsapi.mlb.com/'):
    self.base_url = base_url
    self.log = logging.getLogger(__name__)

  async def execute(self, query: str) -> str:
    team_async = await asyncio.create_task(self.fetch_mlb_team(query))
    team = await team_async
    if not team:
      return 'Not Found'
    roster = await asyncio.create_task(self.fetch_mlb_roster(team["id"]))
    venue = await asyncio.create_task(self.fetch_mlb_venue_by_id(team["venue"]["id"]))
    return self.print_roster(roster, team, venue)

  async def fetch_mlb_roster(self, teamId: int):
    r = requests.get(f'{self.base_url}/api/v1/teams/{teamId}/roster', headers=HEADERS)
    if r.status_code == 200:
      return json.loads(r.content.decode('utf-8'))
    else:
      self.log.error(f'Error fetching roster for {teamId}. Status code: {r.status_code}')
      return None

  async def fetch_mlb_team(self, query: str):
    return self.fetch_mlb_team_by_id(int(query)) if query.isnumeric() \
        else self.fetch_mlb_team_by_name(query)

  async def fetch_mlb_team_by_id(self, teamId: int):
    r = requests.get(f'{self.base_url}/api/v1/teams/{teamId}', headers=HEADERS)
    if r.status_code == 200:
      data = json.loads(r.content.decode('utf-8'))
      return data['teams'][0] if data['teams'] and len(data['teams']) == 1 else None
    else:
      self.log.error(f'Error fetching team for id {teamId}. Status code: {r.status_code}')
      return None

  async def fetch_mlb_team_by_name(self, query: str):
    season = datetime.datetime.now().year
    r = requests.get(
        f'{self.base_url}/api/v1/teams',
        headers=HEADERS,
        params={'season': season, 'sportIds': 1})
    if r.status_code == 200:
      data = json.loads(r.content.decode('utf-8'))
      teams =  [t for t in (data['teams'] or [])
          if t['active'] and query.lower() in t['name'].lower()]
      return teams[0] if len(teams) > 0 else None
    else:
      self.log.error(f'Error fetching roster for query {query}. Status code: {r.status_code}')
      return None

  async def fetch_mlb_venue_by_id(self, venueId: int):
    r = requests.get(f'{self.base_url}/api/v1/venues/{venueId}', headers=HEADERS)
    if r.status_code == 200:
      data = json.loads(r.content.decode('utf-8'))
      venues =  [v for v in (data['venues'] or [])if v['active']]
      return venues[0] if len(venues) > 0 else None
    else:
      self.log.error(f'Error fetching venue for {venueId}. Status code: {r.status_code}')
      return None

  def print_roster(self, roster, team, venue) -> str:
    rows = [['Team', 'Jersey', 'Name', 'Position', 'Home Stadium']]
    players = sorted(roster["roster"], key=lambda p: p["person"]["fullName"])
    for player in roster["roster"]:
      rows.append([
        team['name'],
        player['jerseyNumber'],
        player['person']['fullName'],
        player['position']['abbreviation'],
        venue['name'],
      ])
    csv_buffer = io.StringIO()
    csv_writer = csv.writer(csv_buffer)
    csv_writer.writerows(rows)
    csv_string = csv_buffer.getvalue()
    csv_buffer.close()
    return csv_string
